services/index/src/helpers.js

```diff
@@ -1,111 +1,303 @@
+  fromDb(row) {
+    return {
+      parent: row.parent,
+      name: row.name,
+      expires: row.expires,
+      etag: row.etag,
+    };
+  },
+  // Create a serializable representation of this namespace suitable for response
+  // from an API method.
+  serialize(indexNamespace) {
+    let ns = indexNamespace.parent + '.' + indexNamespace.name;
+    // Remove separate if there is no need
+    if (indexNamespace.parent.length === 0 || indexNamespace.name.length === 0) {
+      ns = indexNamespace.parent + indexNamespace.name;
+    }
+    return {
+      namespace: ns,
+      name: indexNamespace.name,
+      expires: indexNamespace.expires.toJSON(),
+    };
+  },
+  // Call db.get_index_namespaces with named arguments.
+  // You will get a list of up to 1000 entries and a
+  // continuation token.
+  // The response will be of the form { rows, continationToken }.
+  // If there are no namespaces to show, the response will have the
+  // `rows` field set to an empty array.
+  //
+  async getNamespaces(
+    db,
+    { parent, name },
+    {
+      query,
+    } = {},
+  ) {
+    const fetchResults = async (continuation) => {
+      let q = query;
 
-/** Regular expression for valid namespaces */
-exports.namespaceFormat = /^([a-zA-Z0-9_!~*'()%-]+\.)*[a-zA-Z0-9_!~*'()%-]+$/;
+      if (continuation) {
+        q.continuationToken = continuation;
+      }
+
+      const {continuationToken, rows} = await paginateResults({
+        query: q,
+        fetch: (size, offset) => db.fns.get_index_namespaces(
+          parent || parent === '' ? parent : null,
+          name || name === '' ? name : null,
+          size,
+          offset,
+        ),
+      });
+
+      const entries = rows.map(exports.namespaceUtils.fromDb);
+
+      return { rows: entries, continuationToken: continuationToken };
+    };
+
+    // Fetch results
+    return fetchResults(query ? query.continuationToken : {});
+  },
+  /** Create parent structure */
+  async ensureNamespace(db, namespace, expires) {
+    // Stop recursion at root
+    if (namespace.length === 0) {
+      return Promise.resolve(null);
+    }
+
+    // Round to date to avoid updating all the time
+    expires = new Date(
+      expires.getFullYear(),
+      expires.getMonth(),
+      expires.getDate() + 1,
+      0, 0, 0, 0,
+    );
+
+    // Parse namespace
+    if (!(namespace instanceof Array)) {
+      namespace = namespace.split('.');
+    }
+    // Find parent and folder name
+    let name = namespace.pop() || '';
+    let parent = namespace.join('.');
+
+    // Load namespace, to check if it exists and if we should update expires
+    const folder = exports.taskUtils.fromDbRows(await db.fns.get_index_namespace(parent, name));
+
+    if (folder) {
+      // Modify the namespace
+      if (folder.expires < expires) {
+        // Update expires
+        const updatedNamespace = await db.fns.update_index_namespace(parent, name, expires);
+        // Update all parents first though
```

Comment:
I think the parent `ensureNamespace` actually needs to happen before the `update_index_namespace`, as that keeps the invariant that `parent.expires >= child.expires`.